### PR TITLE
eID-Service Integration

### PR DESCRIPTION
**Pull request by BenjaminKeller**
This pull request implements the required functionality of the interfaces of the eID-Service.

It contains the interface to the eID-Server so the SAML-procedure has only to modify the value of `restrictedId` in the `ServiceRequest` table. @Armagetron Please give some feedback about that.

Moreover it provides the view `getUserId` for the openId Provider. The `refreshAddress` provides a json containing the access token:
```
{"accessToken": "9465901b-6693-462b-9044-db3a80fe0cd0"}
```
`getUserId` provides the restricted id as base64:
```
{"userId": "b''"}
```
 @nielsgroth @m273d15 @larissazech Please give some feedback.

See also #44 and #46.

### CODE REVIEW

File: test/e2e/spec/eid_service/main_spec.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    beforeAll(function() {
+        frisby.addExpectHandler('tc token', function(response) {
+            let xml = response._body.replace(/\n/g, "").replace(/\r/g, "").replace(/ /g, "");
+
+            expect(xml).toEqual(jasmine.stringMatching(/^<TCTokenType><ServerAddress>.*<\/Binding><\/TCTokenType>$/));
+        });
+    });
+
+    it("is working fine", function(done) {
```
Comment (nils-wisiol) [inline test/e2e/spec/eid_service/main_spec.js]: The description here should be a little more descriptive. I advise to split this test case into several expectation handlers to make it more readable:

    frisby.get(....getTcTokenUrl....)
      .expect('status', 302)
      .expect('tcToken retrievable')

where the new `tcToken retrievable` expectation handlers implements most of the actual testing.

With #55 in mind, this should be compatible with [chakram](http://dareid.github.io/chakram/jsdoc/module-chakram-expectation.html) by extending [chai](http://chaijs.com/api/bdd/).

File: test/e2e/spec/eid_service/main_spec.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
+describe("eID-Service: The general flow", function() {
```
Comment (nils-wisiol) [inline test/e2e/spec/eid_service/main_spec.js]: This suit just contains one positive test case. Please also add some negative checks to increase the test coverage to the cases where we need to reply faulty requests in the eid service / eid server.

**Review by nils-wisiol — CHANGES_REQUESTED** [review]
As a general point, please increase code documentation throughout your PR.

**Comment by nils-wisiol** [comment]
It may be worthwhile to add a README to the eid server Django app to summarize the interface. Alternatively, this could be entirely put in code docs.

**Comment by BenjaminKeller** [comment]
You're right. Doc is now required as the complexity is high. I put it in code.

**Update:** I added documentation.

TODO:
- [x] Rewrite tests according to new framework
- [x] Add negative tests
- [x] Adapt interface to OpenId-Provider (see #62)

**Comment by auvin** [comment]
Currently the openid provider expects a POST with username, client_id, redirect_uri, response_type, scope and state. The username is used to present the restricted userID, so i think we can than use the provided getUserId for that.
The other attributes are present as soon as the user clicks 'login with eid' and should be forwarded from that point on.

@larissazech, @m273d15 Please correct me, if i am wrong.

Edit: See #62 for a description more precisely.

**Comment by BenjaminKeller** [comment]
Chakram seems to be [broken](https://travis-ci.org/swp-fu-eid/eid-fu-swp/builds/269241584), too?
@nils-wisiol Done.

**Comment by nils-wisiol** [comment]
@BenjaminKeller 
> Chakram seems to be broken, too?

looks good to me, why do you think it is broken?

File: api/eid_service/migrations/0001_initial.py [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # -*- coding: utf-8 -*-
```
Comment (nils-wisiol) [inline api/eid_service/migrations/0001_initial.py]: [This extension of the database model should be in a second file.](https://docs.djangoproject.com/en/1.11/topics/migrations/)

File: api/eid_server/serviceInterface.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from eid_server.models import ServiceRequest
```
Comment (nils-wisiol) [inline api/eid_server/serviceInterface.py]: according to [PEP8](https://www.python.org/dev/peps/pep-0008/), python modules (i.e. source files) should have all lower-case names

**Comment by BenjaminKeller** [comment]
@nils-wisiol The link is not working. Have a look at build #266 in travis. `the tc token generator returns a tc token` is the only test that fails but it affects 6 other tests, too.

Rebasing done.

**Comment by nils-wisiol** [comment]
In the [travis build](https://travis-ci.org/swp-fu-eid/eid-fu-swp/builds/269241584), tests fail as regex do not match. The `Location` that was sent,

    https://eid.local/api/openid/login?eid_access_token=5cd623f4-9cd4-46a4-b49c-ea330fe8fb11&test=bla&foo=bar

does indeed not match the given expression

    /\/api\/openid\/login?eid_access_token=.*[&]test=bla[&]foo=bar$/

Try escaping `?` and `_`:

    /\/api\/openid\/login\?eid\_access\_token=.*[&]test=bla[&]foo=bar$/

[Also useful](http://regexe.de/?ausgangstext=https%3A%2F%2Feid.local%2Fapi%2Fopenid%2Flogin%3Feid_access_token%3D5cd623f4-9cd4-46a4-b49c-ea330fe8fb11%26test%3Dbla%26foo%3Dbar&regexp_text=%5C%2Fapi%5C%2Fopenid%5C%2Flogin%5C%3Feid%5C_access%5C_token%3D.*%5B%26%5Dtest%3Dbla%5B%26%5Dfoo%3Dbar%24&is_dotall=dotall&is_multiline=multiline&replace_text=%241F%242).

**Comment by BenjaminKeller** [comment]
Hi @nils-wisiol your described issue was problem of the test `The refresh page redirects correctly with GET parameters`. And obviously it is now working and I have fixed that a day ago. The other tests (`eID-Service: The tcTokenUrl generator redirects by default` and so on) are in a completely different file with a completely different definition. The regex in this file is also completely different from the displayed. And they fail because the the other test fails. That's the problem.

Nevertheless, this branch can be merged.

**Comment by nils-wisiol** [comment]
I'd blame that problem on the double-definition ([here](https://github.com/swp-fu-eid/eid-fu-swp/blob/2954bfbe13e8820111be75c4b8c729b55471e0c4/test/e2e/spec/eid_service/error_spec.js#L8) and [here](https://github.com/swp-fu-eid/eid-fu-swp/blob/2954bfbe13e8820111be75c4b8c729b55471e0c4/test/e2e/spec/eid_service/refresh_spec.js#L8)) of the `correctRedirect` property.

Chakram (using chai) runs all tests parallel, so you get a race condition and your expectation was overwritten.

I didn't reproduce this locally, but unless there is indication that this theory could be wrong, I'd rather spare the work.

**Review by nils-wisiol — CHANGES_REQUESTED** [review]

**Comment by BenjaminKeller** [comment]
This shouldn't be a problem as every suite has its own `before` call.

Do you have any further suggestions or can we merge?

**Comment by nils-wisiol** [comment]
The double definition is put onto a singleton object, that's why the old one is overwritten. (That's how `require.js` works)

**Comment by BenjaminKeller** [comment]
@nils-wisiol Files renamed, database migrated and tests improved. Do you check line feeds, formatting and style conventions manually or do you have some cool tool?

File: api/eid_service/views.py [hunk 5]
```diff
@@ -67,13 +99,15 @@ def refresh(request):
...
+        authenticationRequest.save()
+        return HttpResponseRedirect('https://' + request.get_host() + '/api/openid/login?eid_access_token=' + str(authenticationRequest.accessToken) + "&" + authenticationRequest.openidParameter)
     else:
         return HttpResponseBadRequest()
-    return HttpResponseBadRequest()
     
 def error(request):
-    return HttpResponse("An error occurred during the authentication.")
+    """
+    The use agent is redirected to this view if the authentication session between eID-Client
```
Comment (nils-wisiol) [inline api/eid_service/views.py]: "user agent"

**Review by nils-wisiol — APPROVED** [review]
This lgtm! Nice work.

I found one more typo, see code annotations. One more general thing, using 'correct' as test description is a tautology: your tests always check if the system behaves correctly. So, if you want to further improve this PR, rename the tests to something avoiding the word 'correct'. How about [adding a method](https://github.com/nils-wisiol/django-rest-boilerplate/blob/20170825_e2e/test/e2e/setup.js#L5-L29) rather than a property that can be used like `expect(...).to.redirect_to(CONST_REGEX)`? Anyways, this is good enough for merging (but please fix the typo real quick -- the approval will stay).

**Comment by nils-wisiol** [comment]
I do not use a tool to check style conventions on reviews (unless I have a suspicion), but I recommend pycharm's integrated style checker. For line endings, I find GNU/Linux' file command helpful. [I used it for checking line endings in a bunch of files at once.](https://github.com/swp-fu-eid/eid-fu-swp/issues/64#issuecomment-325626518)

**Comment by BenjaminKeller** [comment]
@nils-wisiol Thank you very much for your answer. I fixed the typo. I will use a constant redirect test function for testing next time!

**Comment by nils-wisiol** [comment]
Alrighty. Looking forward to see this in master.

On August 29, 2017 10:49:52 PM GMT+02:00, BenjaminKeller <notifications@github.com> wrote:
>@nils-wisiol Thank you very much for your answer. I fixed the typo. I
>will use a constant redirect test function for testing next time!
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/swp-fu-eid/eid-fu-swp/pull/57#issuecomment-325799743

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Pull request merged**